src/function/decoding_function.cpp

```diff
@@ -0,0 +1,145 @@
+                  char *remaining_bytes_buffer, idx_t &remaining_bytes_size) {
+	for (; encoded_buffer_current_position < encoded_buffer_size; encoded_buffer_current_position++) {
+		if (decoded_buffer_current_position == decoded_buffer_size) {
+			// We are done
+			return;
+		}
+		const unsigned char ch = static_cast<unsigned char>(encoded_buffer[encoded_buffer_current_position]);
+		if (ch > 0x7F && ch <= 0x9F) {
+			throw InvalidInputException("File is not latin-1 encoded");
+		}
+		if (ch <= 0x7F) {
+			// ASCII: 1 byte in UTF-8
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(ch);
+		} else {
+			// Non-ASCII: 2 bytes in UTF-8
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(0xc2 + (ch > 0xbf));
+			if (decoded_buffer_current_position == decoded_buffer_size) {
+				// We are done, but we have to store one byte for the next chunk!
+				encoded_buffer_current_position++;
+				remaining_bytes_buffer[0] = static_cast<char>((ch & 0x3f) + 0x80);
+				remaining_bytes_size = 1;
+				return;
+			}
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>((ch & 0x3f) + 0x80);
+		}
+	}
+}
+
+void DecodeUTF8(char *encoded_buffer, idx_t &encoded_buffer_current_position, const idx_t encoded_buffer_size,
+                char *decoded_buffer, idx_t &decoded_buffer_current_position, const idx_t decoded_buffer_size,
+                char *remaining_bytes_buffer, idx_t &remaining_bytes_size) {
+	// This function just copies.
+	for (; encoded_buffer_current_position < encoded_buffer_size; encoded_buffer_current_position++) {
+		if (decoded_buffer_current_position == decoded_buffer_size) {
+			// We are done
+			return;
+		}
+		decoded_buffer[decoded_buffer_current_position++] = encoded_buffer[encoded_buffer_current_position];
+	}
+}
+
+static const DecodingFunction internal_decode_methods[] = {
+    {"utf-8", DecodeLatin1, 1, 0}, {"latin-1", DecodeLatin1, 2, 1}, {"utf-16", DecodeUTF16, 2, 2}};
+
+void DecodingFunctionSet::Initialize(DBConfig &config) {
+	for (auto &decode_method : internal_decode_methods) {
+		config.RegisterDecodeFunction(decode_method);
+	}
+}
+
+void DBConfig::RegisterDecodeFunction(const DecodingFunction &function) const {
+	lock_guard<mutex> l(decoding_functions->lock);
+	const auto decode_type = function.GetType();
+	if (decoding_functions->functions.find(decode_type) != decoding_functions->functions.end()) {
+		throw InvalidInputException("Decoding function with name %s already registered", decode_type);
+	}
+	decoding_functions->functions[decode_type] = function;
+}
+
+optional_ptr<DecodingFunction> DBConfig::GetDecodeFunction(string name) const {
+	lock_guard<mutex> l(decoding_functions->lock);
+	name = StringUtil::Lower(name);
+	// Check if the function is already loaded into the global compression functions.
+	if (decoding_functions->functions.find(name) != decoding_functions->functions.end()) {
```

Comment:
nit: Instead of calling `StringUtil::Lower` - we could turn `decoding_functions->functions` into a `case_insensitive_map`